### PR TITLE
Cortx-31105

### PR DESCRIPTION
**Pull request by rahul-hatwar**
# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by rahul-hatwar** [comment]
{'nodeid': 'tests/s3/versioning/test_versioning_put_object.py::TestVersioningPutObject::test_put_object_preexisting_32724', 'test_id': 'TEST-32724', 'marks': ['s3_ops']}, {'nodeid': 'tests/s3/versioning/test_versioning_put_object.py::TestVersioningPutObject::test_put_object_versioning_enabled_32728', 'test_id': 'TEST-32728', 'marks': ['s3_ops']}, {'nodeid': 'tests/s3/versioning/test_versioning_put_object.py::TestVersioningPutObject::test_put_object_versioning_suspended_32733', 'test_id': 'TEST-32733', 'marks': ['s3_ops']}, {'nodeid': 'tests/security/test_cortx_port_scanner_kubectl_svc.py::test_cortx_port_scanner_kubectl_svc', 'test_id': 'TEST-34217', 'marks': ['security']}, {'nodeid': 'tests/security/test_cortx_port_scanner_netstat.py::test_cortx_port_scanner_netstat', 'test_id': 'TEST-34218', 'marks': ['security']}] created at /root/rahul/cortx-test-1/log/te_meta.json
Successfully unmounted directory

=================================================== 2152 tests collected in 8.37s ===================================================

File: tests/ha/test_control_pod_restart.py [hunk 1]
```diff
@@ -729,3 +731,252 @@ def test_mpu_after_ctrl_pod_failover(self):
...
+                                                    bucket_name=self.bucket_name,
+                                                    object_name=self.object_name,
+                                                    part_numbers=part_numbers,
+                                                    multipart_obj_size=file_size,
+                                                    total_parts=total_parts,
+                                                    multipart_obj_path=self.multipart_obj_path)
+        mpu_id = resp[1]
+        object_path = resp[2]
+        parts_etag1 = resp[3]
+        assert_utils.assert_true(resp[0], f"Failed to upload parts. Response: {resp}")
```
Comment (ankita-93) [inline tests/ha/test_control_pod_restart.py:784]: We should assert first if failure and if not then we should capture resp

File: tests/ha/test_control_pod_restart.py [hunk 2]
```diff
@@ -729,3 +731,252 @@ def test_mpu_after_ctrl_pod_failover(self):
...
+                    self.control_pod_name, failover_node)
+        pod_yaml = {self.control_pod_name: self.modified_yaml}
+        resp = self.ha_obj.failover_pod(pod_obj=self.node_master_list[0], pod_yaml=pod_yaml,
+                                        failover_node=failover_node)
+        assert_utils.assert_true(resp[0], resp)
+        LOGGER.info("Step 3: Successfully failed over control pod to %s. Cluster is in good state",
+                    failover_node)
+
+        self.restore_pod = self.deploy = True
+        LOGGER.info("Step 4: Verify if IAM users %s are persistent across control pod failover",
```
Comment (ankita-93) [inline tests/ha/test_control_pod_restart.py:812]: Please correct the satement as per #1522

File: tests/ha/test_control_pod_restart.py [hunk 3]
```diff
@@ -729,3 +731,252 @@ def test_mpu_after_ctrl_pod_failover(self):
...
+        resp = self.ha_obj.partial_multipart_upload(s3_data=self.s3_clean,
+                                                    bucket_name=self.bucket_name,
+                                                    object_name=self.object_name,
+                                                    part_numbers=remaining_parts,
+                                                    remaining_upload=True, mpu_id=mpu_id,
+                                                    multipart_obj_size=file_size,
+                                                    total_parts=total_parts,
+                                                    multipart_obj_path=object_path)
+
+        assert_utils.assert_true(resp[0], f"Failed to upload parts {resp[1]}")
```
Comment (ankita-93) [inline tests/ha/test_control_pod_restart.py:834]: upload remaining

File: tests/ha/test_control_pod_restart.py [hunk 4]
```diff
@@ -729,3 +731,252 @@ def test_mpu_after_ctrl_pod_failover(self):
...
+        secret_key = resp[1]["secret_key"]
+        s3_test_obj = S3TestLib(access_key=access_key, secret_key=secret_key,
+                                endpoint_url=S3_CFG["s3_url"])
+        uids = [self.s3acc_name]
+        LOGGER.info("Successfully created IAM user with name %s", self.s3acc_name)
+        self.s3_clean = {'s3_acc': {'accesskey': access_key, 'secretkey': secret_key,
+                                    'user_name': self.s3acc_name}}
+
+        LOGGER.info("Step 1: Create and list buckets and perform upload and copy "
+                    "object from %s bucket to other buckets, download objects "
```
Comment (ankita-93) [inline tests/ha/test_control_pod_restart.py:906]: here we not downloading and verifing etags. 
Statement is correct till ....to other buckets.

File: tests/ha/test_control_pod_restart.py [hunk 5]
```diff
@@ -729,3 +731,252 @@ def test_mpu_after_ctrl_pod_failover(self):
...
+                    "and verify etags", self.bucket_name)
+        resp = self.ha_obj.create_bucket_copy_obj(event, s3_test_obj=s3_test_obj,
+                                                  bucket_name=self.bucket_name,
+                                                  object_name=self.object_name,
+                                                  bkt_obj_dict=bkt_obj_dict,
+                                                  file_path=self.multipart_obj_path)
+        assert_utils.assert_true(resp[0], f"failed buckets are: {resp[1]}")
+        put_etag = resp[1]
+        LOGGER.info("Step 1: successfully create and list buckets and perform upload and copy"
+                    "object from %s bucket to other buckets, download objects "
```
Comment (ankita-93) [inline tests/ha/test_control_pod_restart.py:916]: same. till only ____to other buckets.

File: tests/ha/test_control_pod_restart.py [hunk 6]
```diff
@@ -729,3 +731,252 @@ def test_mpu_after_ctrl_pod_failover(self):
...
+        assert_utils.assert_true(resp[0], resp)
+        LOGGER.info("Step 2: Successfully failed over control pod to %s. Cluster is in good state",
+                    failover_node)
+
+        self.restore_pod = self.deploy = True
+        LOGGER.info("Step 3: Verify if IAM users %s are persistent across control pod failover",
+                    uids)
+        for user in uids:
+            resp = self.rest_iam_user.get_iam_user(user)
+            assert_utils.assert_equal(resp.status_code, const.Rest.SUCCESS_STATUS,
```
Comment (ApurwaMohite) [inline tests/ha/test_control_pod_restart.py:940]: same as above.

**Review by ApurwaMohite — APPROVED** [review]

File: tests/ha/test_control_pod_restart.py [hunk 7]
```diff
@@ -1041,3 +1043,249 @@ def test_iam_bkt_cruds_during_ctrl_pod_failover(self):
 
         LOGGER.info("ENDED: Verify IAM user and bucket CRUD operations while control pod is "
                     "failing over")
+    @pytest.mark.ha
```
Comment (ApurwaMohite) [inline tests/ha/test_control_pod_restart.py:1046]: One empty line is required between tests.

**Review by ankita-93 — APPROVED** [review]

File: tests/ha/test_control_pod_restart.py [hunk 8]
```diff
@@ -1041,3 +1043,250 @@ def test_iam_bkt_cruds_during_ctrl_pod_failover(self):
...
+                    self.control_pod_name, failover_node)
+        pod_yaml = {self.control_pod_name: self.modified_yaml}
+        resp = self.ha_obj.failover_pod(pod_obj=self.node_master_list[0], pod_yaml=pod_yaml,
+                                        failover_node=failover_node)
+        assert_utils.assert_true(resp[0], resp)
+        LOGGER.info("Step 2: Successfully failed over control pod to %s. Cluster is in good state",
+                    failover_node)
+
+        self.restore_pod = self.deploy = True
+        LOGGER.info("Step 3: Verify if IAM users %s are persistent across control pod failover",
```
Comment (sonalk0209) [inline tests/ha/test_control_pod_restart.py:1246]: is this a common step for all tests? @ApurwaMohite

File: tests/ha/test_control_pod_restart.py [hunk 9]
```diff
@@ -1041,3 +1043,250 @@ def test_iam_bkt_cruds_during_ctrl_pod_failover(self):
...
+        LOGGER.info("Step 4: Download the uploaded objects & verify etags")
+        for key, val in bkt_obj_dict.items():
+            resp = s3_test_obj.get_object(bucket=key, key=val)
+            LOGGER.info("Get object response: %s", resp)
+            get_etag = resp[1]["ETag"]
+            assert_utils.assert_equal(put_etag, get_etag, "Failed to match GET-PUT ETAG for "
+                                                          f"object {key} of bucket {val}.")
+        LOGGER.info("Step 4: Successfully download the uploaded objects & verify etags")
+
+        if CMN_CFG["dtm0_disabled"]:
```
Comment (sonalk0209) [inline tests/ha/test_control_pod_restart.py:1265]: why this is needed? As we are not putting down any data pod?

File: tests/ha/test_control_pod_restart.py (hunk 8 above)
Comment (ApurwaMohite) [reply tests/ha/test_control_pod_restart.py:1246]: Yes, common in all tests.

Comment (ApurwaMohite) [reply tests/ha/test_control_pod_restart.py:1246]: Yes, common in all tests.

**Review by sonalk0209 — APPROVED** [review]

**Pull request merged**